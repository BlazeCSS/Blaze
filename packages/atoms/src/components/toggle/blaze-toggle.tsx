import { Component, Prop, Method, State } from '@stencil/core';

@Component({
  tag: 'blaze-toggle'
})
export class Toggle {

  @Prop() type: string;
  @Prop() toggled: boolean = false;
  @State() _toggled: boolean = false;

  componentDidLoad() {
    this._toggled = this.toggled;
  }

  @Method()
  isToggled() {
    return this._toggled;
  }

  toggle(e) {
    e.preventDefault();
    this._toggled = !this._toggled;
  }

  render() {
    const type = this.type ? `c-toggle--${this.type}` : '';

    return (
      <label class={`c-toggle ${type}`} onClick={(e) => this.toggle(e)}>
        <input type="checkbox" aria-checked={this._toggled} checked={this._toggled} />
        <div class="c-toggle__track">
          <div class="c-toggle__handle"></div>
        </div>
        <slot />
      </label>
    );
  }
}
