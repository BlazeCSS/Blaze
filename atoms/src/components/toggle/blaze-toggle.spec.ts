import { flush, render } from '@stencil/core/testing';
import { Toggle } from './blaze-toggle';

describe('Toggle', () => {
  it('should build', () => {
    expect(new Toggle()).toBeTruthy();
  });

  let element;

  const snapIt = (name, html) => {
    it(name, async () => {
      element = await render({
        components: [Toggle],
        html
      });
      await flush(element);

      expect(element).toMatchSnapshot();
    });
  };

  snapIt(
    'renders correctly with no props',
    '<blaze-toggle>default</blaze-toggle>'
  );

  snapIt(
    'colours based on type',
    '<blaze-toggle type="success">default</blaze-toggle>'
  );

  snapIt(
    'should be toggle on when prop is present',
    '<blaze-toggle toggled type="success">default</blaze-toggle>'
  );

  it('toggles when clicked on', async () => {
    element = await render({
      components: [Toggle],
      html: '<blaze-toggle>default</blaze-toggle>'
    });
    await flush(element);
    expect(element.isToggled()).toEqual(false);

    element.querySelector('label').click();
    expect(element.isToggled()).toEqual(true);
  });
});
