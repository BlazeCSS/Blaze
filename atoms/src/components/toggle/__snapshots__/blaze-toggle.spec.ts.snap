// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle colours based on type 1`] = `
<blaze-toggle
  class="hydrated"
  type="success"
>
  <label
    class="c-toggle c-toggle--success"
  >
    <input
      type="checkbox"
    />
    <div
      class="c-toggle__track"
    >
      <div
        class="c-toggle__handle"
      />
    </div>
    <!---->
    default
  </label>
</blaze-toggle>
`;

exports[`Toggle renders correctly with no props 1`] = `
<blaze-toggle
  class="hydrated"
>
  <label
    class="c-toggle"
  >
    <input
      type="checkbox"
    />
    <div
      class="c-toggle__track"
    >
      <div
        class="c-toggle__handle"
      />
    </div>
    <!---->
    default
  </label>
</blaze-toggle>
`;

exports[`Toggle should be toggle on when prop is present 1`] = `
<blaze-toggle
  class="hydrated"
  toggled=""
  type="success"
>
  <label
    class="c-toggle c-toggle--success"
  >
    <input
      type="checkbox"
    />
    <div
      class="c-toggle__track"
    >
      <div
        class="c-toggle__handle"
      />
    </div>
    <!---->
    default
  </label>
</blaze-toggle>
`;
